strapi/strapi
# fix: restrict model and attribute names that break Strapi

### What does it do?

Updates the front-end and back-end validator for the CTB for attributes and content types:
- [`then`](https://github.com/strapi/strapi/issues/15557) is a new restricted ct name (which will potentially be removed and allowed in v5)
- [`rest`](https://github.com/strapi/strapi/issues/13643)  is a new restricted ct name which will be [allowed in v5
](https://github.com/strapi/strapi/pull/19365)
- `createdById` and `updatedById` and any casing/snakecase equivalents, are new restricted attribute names -- not a breaking change because using them already broke strapi
- compares with snake_case(itemName) === snake_case(reservedName) since every additional restriction that imposes already prevents strapi from starting -- except the following, which will no longer validate, therefore the CTB will not allowing saving changes to a content type that already contains them:
  - attributes named exactly CreatedBy, UpdatedBy, CREATED_BY, UPDATED_BY
  - user attributes such as MYFIELD and myfield, or some_field and someField, could both be created but only one of them would work


### What does it NOT do?

- validating code schemas on strapi startup to catch problems and point users in the right direction instead of just blocking in the CTB: by far the biggest problem, and this is planned to be released in v4 in the next few weeks (although the degree to which we can do it without breaking changes is still to be seen)
- front-end error handling needs to be improved; it should respect and display the response from the back-end; if FE validation works but BE fails, the error is displayed as generic a 500 message on the front-end
- [`where`](https://github.com/strapi/strapi/issues/10710) appears to be solved already
- [`locale`](https://github.com/strapi/strapi/issues/10181) will not be reserved name in v4 to prevent breaking changes, we will document it and potentially add a warning to Strapi startup; will be fixed in v5
- issues with convert-query-params ([example](https://github.com/strapi/strapi/issues/14511))  many of these seem to be solved and working since posted, but they will be investigated further and definitively fixed in v5 if they still exist
- issues with merging nested data stemming from component, attribute, dz, collection name, etc with the same name ([example](https://github.com/strapi/strapi/issues/11909)) or those nested names being erroneously flagged as non-unique -- we will unfortunately not address these in v4, only v5, due to risk of breaking changes

### Why is it needed?

This and related issues have been a major problem: https://github.com/strapi/strapi/issues/1198

This is just the first step, but fixes are now coming!

### How to test it?

  - the following attributes previously broke strapi to where it wouldn't start, but were accepted by the CTB anyway: `CREATED_BY_ID`, `ID`, `Id`, `createdBy`
    - Note that `iD` should still work for now as it did before
  - the following attributes caused major issues with Strapi even though it would start up: creating both `MY_FIELD` and `my_field` in the same content type
  - the following content type name prevent Strapi from starting up: `then`
  - [Breaking change?] the following attributes were supposed to be reserved, but Strapi worked fine: `CreatedBy`, `UpdatedBy`, `CREATED_BY`, `UPDATED_BY`

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request
